src/snakemake/dag.py

```diff
@@ -1908,6 +1905,22 @@ async def postprocess(
     async def check_needrun_jobs(self):
         for job in filterfalse(self._checked_jobs.__contains__, self.needrun_jobs()):
             await job.check_protected_output()
+            job.is_valid()
+
+            # here we check if no two rules make the same output
+            # this can happen in edge cases where the output of jobs
+            # are not part of the targetfiles
+            for output_file in job.output:
+                if output_file in self._seen_outputs:
+                    other_job = self._seen_outputs[output_file]
+                    if other_job.jobid == job.jobid:
+                        continue
+
+                    if not self.workflow.execution_settings.ignore_ambiguity:
+                        raise AmbiguousRuleException(output_file, other_job, job)
+                else:
+                    self._seen_outputs[output_file] = job
```

Comment:
I think though that this check should happen on all jobs, not only the needrun jobs. Maybe change the name of the method to check_jobs, and have two iterations. One for the needrun jobs, containing check_protected_output, the other one for all jobs, containing is_valid and the new code for the ambiguity detection.